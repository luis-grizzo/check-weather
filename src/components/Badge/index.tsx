interface BadgeProps {
  children: string
}

export function Badge({ children }: BadgeProps) {
  return (
    <span className="flex items-center gap-2 px-2 py-1 text-xs bg-neutral-100/60 border-1 border-neutral-600/10 text-neutral-950 rounded-lg backdrop-blur">
      {children}
    </span>
  )
}
