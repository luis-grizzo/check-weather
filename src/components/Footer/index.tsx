'use client'

export function Footer() {
  const currentYear = new Date().getFullYear()

  return (
    <footer className="bg-neutral-100/60 border-t-1 border-neutral-600/10">
      <p className="container mx-auto px-8 py-6 text-xs text-center">
        Developed by{' '}
        <a
          href="https://luisgrizzo.dev"
          target="_blank"
          rel="noreferrer"
          className="text-blue-600 hover:text-blue-400 transition-colors"
        >
          Lu√≠s Grizzo
        </a>
        {', '}
        <a
          href="https://github.com/luis-grizzo/check-weather?tab=MIT-1-ov-file"
          target="_blank"
          className="text-xs text-blue-600 hover:text-blue-400 transition-colors"
          rel="noreferrer"
        >
          MIT License
        </a>
        {', '}
        {currentYear}.
      </p>
    </footer>
  )
}
