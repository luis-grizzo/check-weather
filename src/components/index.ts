export * from './Badge'
export * from './Button'
export * from './DisplayTime'
export * from './Footer'
export * from './Header'
export * from './InfoPanel'
export * from './Navbar'
export * from './RefreshButton'
