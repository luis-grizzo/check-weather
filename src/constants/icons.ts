import clear_day from '@public/clear_day.svg'
import clear_night from '@public/clear_night.svg'
import partly_cloudy_day from '@public/partly_cloudy_day.svg'
import partly_cloudy_night from '@public/partly_cloudy_night.svg'
import rainy from '@public/rainy.svg'
import thunderstorm from '@public/thunderstorm.svg'
import weather_snowy from '@public/weather_snowy.svg'
import mist from '@public/mist.svg'
import weather_mix from '@public/weather_mix.svg'
import foggy from '@public/foggy.svg'
import airwave from '@public/airwave.svg'
import storm from '@public/storm.svg'
import tornado from '@public/tornado.svg'

import refresh from '@public/refresh.svg'
import cloud_off from '@public/cloud_off.svg'
import air from '@public/air.svg'
import visibility from '@public/visibility.svg'
import water_drop from '@public/water_drop.svg'
import info from '@public/info.svg'
import error from '@public/error.svg'
import progress_activity from '@public/progress_activity.svg'

export const weatherIcons = {
  clear_day: {
    src: clear_day,
    alt: 'Clear day icon'
  },
  clear_night: {
    src: clear_night,
    alt: 'Clear night icon'
  },
  partly_cloudy_day: {
    src: partly_cloudy_day,
    alt: 'Partly cloudy day icon'
  },
  partly_cloudy_night: {
    src: partly_cloudy_night,
    alt: 'Partly cloudy night icon'
  },
  rainy: {
    src: rainy,
    alt: 'Rainy icon'
  },
  thunderstorm: {
    src: thunderstorm,
    alt: 'Thunderstorm icon'
  },
  weather_snowy: {
    src: weather_snowy,
    alt: 'Weather snowy icon'
  },
  mist: {
    src: mist,
    alt: 'Mist icon'
  },
  weather_mix: {
    src: weather_mix,
    alt: 'Weather mix icon'
  },
  foggy: {
    src: foggy,
    alt: 'Foggy icon'
  },
  airwave: {
    src: airwave,
    alt: 'Airwave icon'
  },
  storm: {
    src: storm,
    alt: 'Storm icon'
  },
  tornado: {
    src: tornado,
    alt: 'Tornado icon'
  }
}

export const layoutIcons = {
  refresh,
  cloud_off,
  air,
  visibility,
  water_drop,
  info,
  error,
  progress_activity
}

export const weatherIconsArray = Object.values(weatherIcons)
