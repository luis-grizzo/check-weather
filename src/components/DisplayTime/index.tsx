'use client'

import { Badge } from '@/components/Badge'

import { formatTime } from '@/lib/stringFormaters'

interface DisplayTimeProps {
  description: string
  timestamp: number
}

export function DisplayTime({ description, timestamp }: DisplayTimeProps) {
  return <Badge>{`${description} ${formatTime(timestamp)}`}</Badge>
}
