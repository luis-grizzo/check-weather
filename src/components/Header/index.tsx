'use client'

import { useEffect, useState } from 'react'

import { hourInMilliseconds } from '@/constants/timeMarks'

import {
  formatCountryName,
  formatDate,
  formatTime,
  formatTimer
} from '@/lib/stringFormaters'
import { positiveOrZero } from '@/lib/numberFormaters'

import type { FetchWeatherFactoredResponse } from '@/utils/weatherUtils'
import { useRouter } from 'next/navigation'

interface HeaderProps {
  location: FetchWeatherFactoredResponse['location']
  requestTimestamp: number
}

export function Header({ location, requestTimestamp }: HeaderProps) {
  const router = useRouter()

  const [countdownTimer, setCountdownTimer] = useState('--:--')
  const [shouldRefresh, setShouldRefresh] = useState(false)

  useEffect(() => {
    const timer = setInterval(() => {
      const currentTime = Date.now()
      const revalidateTime = requestTimestamp * 1_000 + hourInMilliseconds

      const remainTime = revalidateTime - currentTime
      const validatedRemainTime = positiveOrZero(remainTime)

      const formattedTimer = formatTimer(validatedRemainTime)

      setCountdownTimer(formattedTimer)

      if (!validatedRemainTime) {
        setShouldRefresh(true)

        clearInterval(timer)
      }
    }, 1_000)

    return () => clearInterval(timer)
  }, [requestTimestamp])

  useEffect(() => {
    if (shouldRefresh) {
      router.refresh()

      setShouldRefresh(false)
    }
  }, [shouldRefresh])

  return (
    <header className="flex items-center justify-between container mx-auto px-8 py-6">
      <div className="flex flex-col gap-1">
        <span role="timer" className="text-xs">
          {`Next forecast in ${countdownTimer}`}
        </span>

        {location.country && (
          <span className="text-lg">
            {`${location.city}, ${formatCountryName(location.country)}`}
          </span>
        )}

        <span className="text-xs">{`${formatTime(requestTimestamp)} - ${formatDate(requestTimestamp)}`}</span>
      </div>
    </header>
  )
}
