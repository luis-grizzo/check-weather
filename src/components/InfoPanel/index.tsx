import Image from 'next/image'

import { layoutIcons } from '@/constants/icons'

interface InfoPanelProps {
  children: React.ReactNode
}

export function InfoPanel({ children }: InfoPanelProps) {
  return (
    <div
      role="dialog"
      className="flex items-start gap-4 p-4 text-sm text-red-950 bg-red-100/60 border-red-600/10 border-1 rounded-xl"
    >
      <Image src={layoutIcons.error} alt="" className="w-5 aspect-square" />

      {children}
    </div>
  )
}
