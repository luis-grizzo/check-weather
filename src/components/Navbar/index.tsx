'use client'

import { useEffect, useState } from 'react'
import Image from 'next/image'
import { motion, AnimatePresence } from 'framer-motion'

import { weatherIconsArray } from '@/constants/icons'
import { calculateRandomIndex } from '@/lib/calculators'

export function Navbar() {
  const [stateIndex, setStateIndex] = useState(0)

  const MotionImage = motion(Image)

  useEffect(() => {
    const timer = setTimeout(() => {
      let newIndex = stateIndex

      while (newIndex === stateIndex)
        newIndex = calculateRandomIndex(weatherIconsArray)

      setStateIndex(newIndex)
    }, 5_000)

    return () => clearTimeout(timer)
  }, [stateIndex])

  return (
    <nav className="sticky top-0 min-h-[76px] bg-neutral-100/80 border-b-1 border-neutral-600/10 backdrop-blur overflow-hidden z-50">
      <div className="flex items-center gap-2 container mx-auto px-8 py-6">
        <AnimatePresence mode="wait">
          {weatherIconsArray.map(
            (icon, index) =>
              stateIndex === index && (
                <MotionImage
                  key={icon.alt}
                  role="img"
                  initial={{ opacity: 0, y: -32 }}
                  animate={{ opacity: 1, y: 0 }}
                  exit={{ opacity: 0, y: 32 }}
                  src={icon.src}
                  alt={icon.alt}
                  className="w-7 aspect-square"
                />
              )
          )}
        </AnimatePresence>

        <span className="text-lg">Check Weather</span>
      </div>
    </nav>
  )
}
