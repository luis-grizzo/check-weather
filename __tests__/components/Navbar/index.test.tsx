import { act } from 'react'
import { render, screen, waitFor } from '@testing-library/react'

import { weatherIconsArray } from '@/constants/icons'

import { Navbar } from '@/components/Navbar'

jest.useFakeTimers()

describe('Navbar', () => {
  beforeEach(() => {
    render(<Navbar />)
  })

  it('should match snapshot', () => {
    const { container } = render(<Navbar />)

    expect(container).toMatchSnapshot()
  })

  it('should change the icon every 5 seconds', async () => {
    const possiblyAltTexts = weatherIconsArray.map((icon) => icon.alt)

    act(() => {
      jest.advanceTimersByTime(5_000)
    })

    await waitFor(() =>
      expect(
        possiblyAltTexts.includes(
          (screen.getByRole('img') as HTMLImageElement).alt
        )
      ).toBeTruthy()
    )
  })
})
