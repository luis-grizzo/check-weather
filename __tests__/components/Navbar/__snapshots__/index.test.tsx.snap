// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should match snapshot 1`] = `
<div>
  <nav
    class="sticky top-0 min-h-[76px] bg-neutral-100/80 border-b-1 border-neutral-600/10 backdrop-blur overflow-hidden z-50"
  >
    <div
      class="flex items-center gap-2 container mx-auto px-8 py-6"
    >
      <img
        alt="Clear day icon"
        class="w-7 aspect-square"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        role="img"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent; opacity: 0; transform: translateY(-32px) translateZ(0);"
        width="40"
      />
      <span
        class="text-lg"
      >
        Check Weather
      </span>
    </div>
  </nav>
</div>
`;
